### PR TITLE
Add ZoneGrid component

### PR DESCRIPTION
**Pull request by TomTirapani**
See Toolbox PR: https://github.com/xh/toolbox/pull/679

Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [x] Added CHANGELOG entry, or determined not required.
- [x] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [x] Updated doc comments / prop-types, or determined not required.
- [x] Reviewed and tested on Mobile, or determined not required.
- [x] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

File: mobile/cmp/button/MultiZoneMapperButton.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ */
+import {hoistCmp, useContextModel} from '@xh/hoist/core';
+import {MultiZoneGridModel} from '@xh/hoist/mobile/cmp/multiZoneGrid';
+import {button, ButtonProps} from '@xh/hoist/mobile/cmp/button';
+import {Icon} from '@xh/hoist/icon';
+import {withDefault} from '@xh/hoist/utils/js';
+import '@xh/hoist/mobile/register';
+
+export interface MultiZoneMapperButtonProps extends ButtonProps {
+    multiZoneGridModel?: MultiZoneGridModel;
```
Comment (lbwexler) [inline mobile/cmp/button/MultiZoneMapperButton.ts:15]: nit: wonder if this could just be `gridModel`.  Let the type do the work

Comment (lbwexler) [reply mobile/cmp/button/MultiZoneMapperButton.ts:15]: ah -- see it is actually not a `GridModel` superclass. Not sure about this.

File: mobile/cmp/multiZoneGrid/MultiZoneGrid.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * A MultiZoneGrid is a specialized version of the Grid component.
+ *
+ * It displays its data with multi-line full-width rows, each broken into four zones for
+ * top/bottom and left/right - (tl, tr, bl, br). Zone mappings determine which of the
+ * available fields should be extracted from the record and rendered into each zone.
+ */
+export const [MultiZoneGrid, multiZoneGrid] = hoistCmp.withFactory<MultiZoneGridProps>({
+    displayName: 'MultiZoneGrid',
```
Comment (lbwexler) [inline mobile/cmp/multiZoneGrid/MultiZoneGrid.ts:40]: We could consider calling this `ZonedGrid` everywhere  if we wanted something shorter, with two words only...

File: mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+    leftColumnSpec?: Partial<ColumnSpec>;
+
+    /** Optional configs to apply to right column */
+    rightColumnSpec?: Partial<ColumnSpec>;
+
+    /** String rendered between consecutive SubFields. */
+    delimiter?: string;
+
+    /** Config with which to create a MultiZoneMapperModel, or boolean `true` to enable default. */
+    multiZoneMapperModel?: MultiZoneMapperConfig | boolean;
```
Comment (lbwexler) [inline mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts:51]: this name definitely benefits from removal of the word 'multi'

File: mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+
+        const {mappings, sortBy, groupBy} = this._defaultState;
+        this.setMappings(mappings);
+        this.setSortBy(sortBy);
+        this.setGroupBy(groupBy);
+
+        this.persistenceModel?.clear();
+
+        if (this.restoreDefaultsFn) {
+            await this.restoreDefaultsFn();
```
Comment (lbwexler) [inline mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts:183]: use conditional eval on 183 to match line  180?

File: mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+            // Ensure all available columns are provided as hidden columns for lookup by multifield renderer
+            ...this.availableColumns
+        ];
+    }
+
+    private buildMultiZoneColumn(isLeft: boolean): ColumnSpec {
+        const topMappings = this.mappings[isLeft ? 'tl' : 'tr'],
+            bottomMappings = this.mappings[isLeft ? 'bl' : 'br'];
+
+        if (isEmpty(topMappings)) {
```
Comment (lbwexler) [inline mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts:232]: throwIf

File: mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+     */
+    columns: Array<ColumnSpec>;
+
+    /** Mappings of columns to zones. */
+    mappings: Record<Zone, Some<string | ZoneMapping>>;
+
+    /** Optional configurations for zone constraints. */
+    limits?: Partial<Record<Zone, ZoneLimit>>;
+
+    /** Optional configs to apply to left column */
```
Comment (lbwexler) [inline mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts:41]: Should we specify that this is an escape hatch? As specified, setting many of these will break stuff pretty badly

File: mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+            absSort: primaryCol.absSort,
+            sortingOrder: primaryCol.sortingOrder,
+            sortValue: primaryCol.sortValue,
+            sortToBottom: primaryCol.sortToBottom,
+            comparator: primaryCol.comparator,
+            sortable: primaryCol.sortable,
+            getValueFn: primaryCol.getValueFn,
+
+            // Optional overrides
+            ...(isLeft ? this.leftColumnSpec : this.rightColumnSpec)
```
Comment (lbwexler) [inline mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts:279]: wondering if the `xxxColumnSpec` should come *before* the primaryCol settings. Could be a more useful default like that, and the columns are also settable by end user...

File: mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+
+                const ret = isString(it) ? {field: it} : it,
+                    col = this.findColumnSpec(ret);
+
+                if (!col) throw XH.exception(`Column not found for field ${ret.field}`);
+                return mapping.push(ret);
+            });
+
+            // 2) Ensure mapping respects configured limits
+            const limit = this.limits?.[zone];
```
Comment (lbwexler) [inline mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts:308]: I wonder about 'constraints' vs.  'limits'

File: mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+            if (limit) {
+                if (isFinite(limit.min) && mapping.length < limit.min) {
+                    throw XH.exception(`Requires minimum ${limit.min} mappings in zone "${zone}"`);
```
Comment (lbwexler) [inline mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts:311]: use throwIf throughout method

File: mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+        if (isPlainObject(mapperModel)) {
+            return new MultiZoneMapperModel({
+                ...(mapperModel as MultiZoneMapperConfig),
+                multiZoneGridModel: this
+            });
+        }
+        return mapperModel ? new MultiZoneMapperModel({multiZoneGridModel: this}) : null;
+    }
+
+    //-----------------------
```
Comment (lbwexler) [inline mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts:340]: probably worth putting these above implemenation block

File: mobile/cmp/multiZoneGrid/Types.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    showLabel?: boolean;
+}
+
+export interface ZoneLimit {
+    /** Min number of fields that should be mapped to the zone */
+    min?: number;
+    /** Max number of fields that should be mapped to the zone */
+    max?: number;
+    /** Array of allowed fields for the zone */
+    only?: string[];
```
Comment (lbwexler) [inline mobile/cmp/multiZoneGrid/Types.ts:25]: `allowedFields` ?

File: mobile/cmp/multiZoneGrid/impl/MultiZoneMapper.ts [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+/**
+ * Hoist UI for user selection and discovery of available MultiZoneGrid columns, enabled via the
+ * `MultiZoneGridModel.multiZoneMapperModel` config option.
+ *
+ * This component displays an example of each of the four zones, with the available columns for
+ * the currently selected zone displayed in a list below. Users can toggle column visibility
+ * and labels for each zone to construct a custom layout for the grid rows.
+ *
+ * It is not necessary to manually create instances of this component within an application.
+ */
```
Comment (lbwexler) [inline mobile/cmp/multiZoneGrid/impl/MultiZoneMapper.ts:28]: Mark with @internal

File: mobile/cmp/multiZoneGrid/Types.ts (hunk 11 above)
Comment (TomTirapani) [reply mobile/cmp/multiZoneGrid/Types.ts:25]: Think its a better name, and happy to change, but wanted to point out that this name (and 'limits') were taken from the Sencha API.

File: mobile/cmp/multiZoneGrid/impl/MultiZoneMapperModel.ts [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+        this.gridModel = this.createGridModel();
+
+        this.addReaction(
+            {
+                track: () => this.selectedZone,
+                run: () => this.syncMapperDataForZone()
+            },
+            {
+                track: () => XH.routerState,
+                run: this.close
```
Comment (lbwexler) [inline mobile/cmp/multiZoneGrid/impl/MultiZoneMapperModel.ts:100]: nit: Line 100 and Line 96 should both be either dynamically bound, or both be fat arrow functions

File: mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts (hunk 4 above)
Comment (TomTirapani) [reply mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts]: I tried that, but IJ complained about it with the `await`

File: mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts (hunk 7 above)
Comment (TomTirapani) [reply mobile/cmp/multiZoneGrid/MultiZoneGridModel.ts]: The main intention here is to allow the user to set their own column flex, which was managed with dedicated properties over in the Sencha version. I was trying to be flexible here, but maybe its better to lock it down to just offering dedicated flex props?

File: mobile/cmp/button/MultiZoneMapperButton.ts (hunk 1 above)
Comment (TomTirapani) [reply mobile/cmp/button/MultiZoneMapperButton.ts]: I renamed to `zonedGridModel?: ZonedGridModel`

File: cmp/zonedGrid/Types.ts [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    displayName: string;
+    label: string;
+    renderer: ColumnRenderer;
+    column: Column;
+    chooserGroup: string;
+    sortable: boolean;
+    sortingOrder: ColumnSortSpec[];
+}
+
+export interface ZonedGridModelPersistOptions extends PersistOptions {
```
Comment (lbwexler) [inline cmp/zonedGrid/Types.ts:40]: wonder if we should go with `ZoneGrid` to match `ZoneMapper` , `Zone` etc.

File: cmp/zonedGrid/Types.ts [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+export interface ZoneLimit {
+    /** Min number of fields that should be mapped to the zone */
+    min?: number;
+    /** Max number of fields that should be mapped to the zone */
+    max?: number;
+    /** Array of allowed fields for the zone */
+    only?: string[];
+}
+
+export interface MapperField {
```
Comment (lbwexler) [inline cmp/zonedGrid/Types.ts:29]: `ZoneField`  might tie this all together with the common prefix

File: cmp/zonedGrid/ZoneMapperModel.ts [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+    /** True (default) to group columns by their chooserGroup */
+    groupColumns?: boolean;
+}
+
+/**
+ * State management for the ZoneMapper component.
+ *
+ * It is not necessary to manually create instances of this class within an application.
+ *
+ * @internal
```
Comment (lbwexler) [inline cmp/zonedGrid/ZoneMapperModel.ts:34]: if truly internal (and I think it is) should be in `/impl`

File: utils/js/LangUtils.ts [hunk 17]
```diff
@@ -283,6 +284,15 @@ export function filterConsecutive<T>(
     };
 }
 
+/**
+ * Intersperse a separator between each item in an array.
+ */
+export function intersperse<T>(arr: T[], separator: any): T[] {
```
Comment (lbwexler) [inline utils/js/LangUtils.ts:290]: nifty!  I have always wanted such a thing.  Not sure about the type.  Seems like Seperator should be also of type `T`.  Or alternatively,  could have another parameter `S` and have it return `Array<T|S>`

File: desktop/cmp/button/ZoneMapperButton.ts [hunk 18]
```diff
@@ -69,11 +71,18 @@ export const [ZoneMapperButton, zoneMapperButton] = hoistCmp.withFactory<ZoneMap
                     zoneMapper({model: mapperModel})
                 ]
             }),
-            onInteraction: willOpen => {
-                if (willOpen) {
+            onInteraction: (willOpen, e?) => {
+                if (isOpen && !willOpen) {
+                    // Prevent clicks with Select controls from closing popover
+                    const selectPortal = document.getElementById(MENU_PORTAL_ID),
```
Comment (lbwexler) [inline desktop/cmp/button/ZoneMapperButton.ts:77]: yikes --  this could hit any application.  Is there a way we can change select to not leak these?

Comment (TomTirapani) [reply desktop/cmp/button/ZoneMapperButton.ts:77]: Yeah, we have similar patterns in some client apps that would be great to unwind in favour of built-in protection. I'll file a ticket 👍

**Pull request merged**